consensus/src/pos/target.rs

```diff
@@ -0,0 +1,477 @@
+        fn get_block_index(
+            &self,
+            block_id: &Id<Block>,
+        ) -> Result<Option<BlockIndex>, PropertyQueryError> {
+            Ok(self.blocks.iter().find(|(id, _)| id == block_id).map(|(_, b)| b.clone()))
+        }
+
+        fn get_gen_block_index(
+            &self,
+            block_id: &Id<GenBlock>,
+        ) -> Result<Option<GenBlockIndex>, PropertyQueryError> {
+            match block_id.classify(self.chain_config) {
+                GenBlockId::Genesis(_) => Ok(Some(GenBlockIndex::Genesis(Arc::clone(
+                    self.chain_config.genesis_block(),
+                )))),
+                GenBlockId::Block(id) => Ok(self
+                    .blocks
+                    .iter()
+                    .find(|(current_id, _)| *current_id == id)
+                    .map(|(_, b)| b.clone().into_gen_block_index())),
+            }
+        }
+
+        fn get_ancestor(
+            &self,
+            _block_index: &BlockIndex,
+            _ancestor_height: common::primitives::BlockHeight,
+        ) -> Result<GenBlockIndex, PropertyQueryError> {
+            unimplemented!()
+        }
+
+        fn get_block_reward(
+            &self,
+            _block_index: &BlockIndex,
+        ) -> Result<Option<common::chain::block::BlockReward>, PropertyQueryError> {
+            unimplemented!()
+        }
+
+        fn get_epoch_data(
+            &self,
+            _epoch_index: u64,
+        ) -> Result<Option<chainstate_types::EpochData>, PropertyQueryError> {
+            unimplemented!()
+        }
+    }
+
+    #[rstest]
+    #[trace]
+    #[case(Seed::from_entropy())]
+    fn calculate_new_target_test(#[case] seed: Seed) {
+        let mut rng = make_seedable_rng(seed);
+        let config = create_unittest_pos_config();
+        {
+            // average block time <= target block time
+            let prev_target = Uint256::from_u64(rng.gen::<u64>());
+            let average_block_time = config.target_block_time().as_secs() / rng.gen_range(1..10);
+            let new_target =
+                calculate_new_target(&config, &prev_target, average_block_time).unwrap();
+            assert!(new_target <= Compact::from(prev_target));
+        }
+        {
+            // average block time >= target block time
+            let prev_target = Uint256::from_u64(rng.gen::<u64>());
+            let average_block_time = config.target_block_time().as_secs() * rng.gen_range(1..10);
+            let new_target =
+                calculate_new_target(&config, &prev_target, average_block_time).unwrap();
+            assert!(new_target >= Compact::from(prev_target));
+        }
+    }
+
+    #[rstest]
+    #[trace]
+    #[case(Seed::from_entropy())]
+    fn calculate_new_target_too_easy(#[case] seed: Seed) {
+        let mut rng = make_seedable_rng(seed);
+        let config = PoSChainConfig::new(Uint256::ONE, Duration::from_secs(1), 1.into(), 1);
+        let prev_target = rng.gen_range(2..u64::MAX);
```

Comment:
Why not random Uint256? You can easily do that by going from H256.

Reply:
I want it to be always bigger than initial target which is set to 1